pczt/src/sapling.rs

```diff
@@ -0,0 +1,548 @@
+    /// "None", to make the output unrecoverable from the chain by the sender.
+    pub(crate) ock: Option<[u8; 32]>,
+
+    /// The ZIP 32 derivation path at which the spending key can be found for the output.
+    pub(crate) zip32_derivation: Option<Zip32Derivation>,
+
+    /// Proprietary fields related to the note being spent.
+    #[getset(get = "pub")]
+    pub(crate) proprietary: BTreeMap<String, Vec<u8>>,
+}
+
+impl Bundle {
+    /// Merges this bundle with another.
+    ///
+    /// Returns `None` if the bundles have conflicting data.
+    pub(crate) fn merge(mut self, other: Self) -> Option<Self> {
+        // Destructure `other` to ensure we handle everything.
+        let Self {
+            mut spends,
+            mut outputs,
+            value_sum,
+            anchor,
+            bsk,
+        } = other;
+
+        // If `bsk` is set on either bundle, the IO Finalizer has run, which means we
+        // cannot have differing numbers of spends or outputs, and the value balances must
+        // match.
+        match (self.bsk.as_mut(), bsk) {
+            (Some(lhs), Some(rhs)) if lhs != &rhs => return None,
+            (Some(_), _) | (_, Some(_))
+                if self.spends.len() != spends.len()
+                    || self.outputs.len() != outputs.len()
+                    || self.value_sum != value_sum =>
+            {
+                return None
+            }
+            // IO Finalizer has run, and neither bundle has excess spends or outputs.
+            (Some(_), _) | (_, Some(_)) => (),
+            // IO Finalizer has not run on either bundle. If the other bundle has more
+            // spends or outputs than us, move them over; these cannot conflict by
+            // construction.
+            (None, None) => {
+                let (other_has_more_spends, other_has_more_outputs) = match (
+                    spends.len().cmp(&self.spends.len()),
+                    outputs.len().cmp(&self.outputs.len()),
+                ) {
+                    // These cases require us to recalculate the value sum, which we can't
+                    // do without a parsed bundle.
+                    (Ordering::Less, Ordering::Greater) | (Ordering::Greater, Ordering::Less) => {
+                        return None
+                    }
+                    // These cases mean that at least one of the two value sums is correct
+                    // and we can use it directly.
+                    (spends, outputs) => (
+                        matches!(spends, Ordering::Greater),
+                        matches!(outputs, Ordering::Greater),
+                    ),
+                };
+
+                if other_has_more_spends {
+                    self.spends.extend(spends.drain(self.spends.len()..));
+                }
+                if other_has_more_outputs {
+                    self.outputs.extend(outputs.drain(self.outputs.len()..));
+                }
```

Comment:
This does not check whether we have any signatures (transparent or spend auth sigs). If we do then modifying either `self.spends` or `self.outputs` would invalidate them.